harness/determined/pytorch/_metric_utils.py

```diff
@@ -0,0 +1,162 @@
+from typing import Any, Dict, List, Optional, Tuple, Union, cast
+
+import numpy as np
+import torch
+
+import determined as det
+from determined import pytorch, util
+
+
+def _combine_metrics_across_processes(
+    context: det._core.DistributedContext, metrics: Dict[str, Any], num_batches: int
+) -> Tuple[Optional[Dict[str, Any]], Optional[List[int]], Optional[int]]:
+    # The chief receives the metric from every other training process.
+    assert (
+        context.size > 1
+    ), "_combine_metrics_across_processes should only be called if context.distributed > 1"
+
+    # all_args is a list of [(metrics, num_batches), ...] for each worker.
+    all_args = context._zmq_gather((metrics, num_batches))
+
+    if not context.rank == 0:
+        return None, None, None
+
+    # Remove items without keys in dictionary. These are from intermediate model parallel nodes.
+    all_args = [a for a in cast(List, all_args) if len(a[0])]
```

Comment:
rather than cast here, can you just do an `assert all_args is not None` above?  It should be true after `if not context.rank == 0: return` but mypy needs that help.

Reply:
(the big difference between a cast and an assert is that the assert will alert you if your assumption ever becomes untrue)